### PR TITLE
Add errorIfMissingOrEmpty attribute to ChangeLogProperty and ChangeLogInclude classes

### PR DESCRIPTION
**Pull request by jasonlyle88**
## Impact

- [X] Bug fix (non-breaking change which fixes expected existing functionality)
- [X] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Add the `errorIfMissingOrEmpty` functionality that is available in `includeAll` to `ChangeLogInclude` and `ChangeLogProperty` classes (when `property` is using a file).

## Things to be aware of

I have set the default value for `errorIfMissingOrEmpty` to be `true` for both `property` and `include`. This is the functionality for `include` and was the functionality for `property` until PR https://github.com/liquibase/liquibase/pull/3595. So this PR adds the `errorIfMissingOrEmpty` functionality as well as fixes the change from original behavior introduced in PR https://github.com/liquibase/liquibase/pull/3595. 

Additionally I have created tests for all 18 possible use cases that could arise with a property file based on 3 possible values for `errorIfMissingOrEmpty` (unset, `true`, `false`) + the 3 possible values for `relativeToChangelogFile` (unset, `true`, `false`) + 2 categorical values for `file` (A value that can be resolved to a file and an value that cannot be resolved to a file). I believe that should cover everything, but let me know if not.

I have also added two additional tests for `include` and the `errorIfMissingOrEmpty` attribute.

## Things to worry about

Nothing of which I am aware.

## Additional Context

@filipelautert This is the change you asked me to tag you in from PR https://github.com/liquibase/liquibase/pull/3595

### CODE REVIEW

**Comment by jasonlyle88** [comment]
Since the original PR (`errorIfMissingOrEmpty` for only `ChangeLogProperty`), I realized I forgot that I wanted to implement this functionality for both `ChangeLogProperty` and ``ChangeLogInclude`. So thats what the update of the title and original description are for (as well as the new commit)

**Review by nvoxland — COMMENTED** [review]
The `var/folders` files should not have been committed

The new files in `test/java/liquibase/verify` are ones that could be committed since they get generated, but maybe best to leave out of this PR since I'd expect nothing in those directories to be impacted and the newly committed files makes it harder to see that.

Would you be able to clear those out, @jasonlyle88 ?

**Comment by jasonlyle88** [comment]
@nvoxland Absolutely, easy enough to do. Do you want me to add the following entries to the .gitignore file as well:

1. liquibase-core/src/test/java/liquibase/verify/**
2. liquibase-core/var/folders/**

Or maybe:

1. */src/test/java/liquibase/verify/**
2. */var/folders/**

?

Just let me know

**Comment by nvoxland** [comment]
Thanks for fixing it up, @jasonlyle88

> Do you want me to add the following entries to the .gitignore file as well?

We don't want the "verify" one in there. We want to be tracking files in there, there's just some that haven't been added like they should be, so the fix is to bring those in as an independent PR.

The "var" ones shouldn't be added ever, but I'm not sure where those came from. Maybe something unique to your setup? If that's the case, we shouldn't add it. If it was created by normal test running or something then we should consider it.

**Comment by jasonlyle88** [comment]
@nvoxland Makes sense. I'm using IntelliJ Idea. The "var" files showed up when I ran the TEST maven lifecycle, nothing special as far as I'm concerned. Happy to add the `*/var/folders/**` to the gitignore if you want, or I can just make sure not to add them for now... either way works for me!

**Review by MalloD12 — APPROVED** [review]
**Approved**.

## Review and testing results:

Thank you @jasonlyle88 for another PR. Overall this one looks good to me. I have made few minor changes which I describe below:
- Update some `DatabaseChanageLog` tests which were failing after resolving some merge conflicts.
- Rename attribute name from `errorIfMissingOrEmpty` to `errorIfMissing`. I'll leave this open for discussion, but I basically consider this change needed as we are not adding any extra logic to check and do some action if change log is empty.
- Add an error exception in case `includeAll()` fails when trying to read change logs from specified directory. 
- Tests successfully executed.

### Things to be aware of:
- None

### Things to worry about:
- None

**Comment by jasonlyle88** [comment]
Hey @MalloD12 , no problem, thanks for taking care of everything for me again!

Only issue I have is that I never touched the `includeAll` stuff. That is all stuff that has been around for a bit, so I'm not sure if you want to change the `includeAll` from `errorIfMissingOrEmpty` to `errorIfMissing` for this specific case. I think `includeAll` does specific checks for empty directories. My stuff should absolutely be changed to `errorIfMissing` since it doesn't do anything special if the specified file is empty.

**Comment by jasonlyle88** [comment]
I guess my concern is it is existing functionality for `includeAll` and this is a breaking change.  So anybody that has changelogs already utilizing the `errorIfMissingOrEmpty` attribute of an `includeAll` element will all of the sudden have their changelog fail

**Comment by MalloD12** [comment]
> I guess my concern is it is existing functionality for `includeAll` and this is a breaking change. So anybody that has changelogs already utilizing the `errorIfMissingOrEmpty` attribute of an `includeAll` element will all of the sudden have their changelog fail

Yeap, this is changing a bit the logic I shared above in the case `errorIfMissingOrEmpty = false`. I think it shouldn't be a `&&` and we should change it to  an `||`, but doing that I'm not really sure whether it makes too much sense about having this one attribute for `includeAll`. What do you guys think about it @filipelautert @XDelphiGrl?

**Comment by jasonlyle88** [comment]
My wife is going to be having our first baby any day now, so I'll probably be absent for a while! I'm good with everything here, just wanted to raise the points above to make sure they were recognized. Look forward to seeing this merged in when I eventually emerge from a daze of no sleep 🤣

**Comment by XDelphiGrl** [comment]
Congratulations, @jasonlyle88!

**Comment by MalloD12** [comment]
> My wife is going to be having our first baby any day now, so I'll probably be absent for a while! I'm good with everything here, just wanted to raise the points above to make sure they were recognized. Look forward to seeing this merged in when I eventually emerge from a daze of no sleep 🤣

Congratulations @jasonlyle88, wish you all the best!

**Review by filipelautert — APPROVED** [review]

**Comment by XDelphiGrl** [comment]
Functional Tests Rerunning Here: [https://github.com/liquibase/liquibase-pro-tests/actions/runs/5003360452](https://github.com/liquibase/liquibase-pro-tests/actions/runs/5003360452)

Test Harness Execution Running Here: [https://github.com/liquibase/liquibase-test-harness/actions/runs/5003369502](https://github.com/liquibase/liquibase-test-harness/actions/runs/5003369502)

File: liquibase-standard/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.21.xsd
Comment (XDelphiGrl) [inline liquibase-standard/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.21.xsd:1]: @filipelautert , hi - I believe we need to make the change in `dbchangelog-latest.xsd`, not in the 4.21 version of the XSD.

**Comment by XDelphiGrl** [comment]
Added `releaseMajor` label as there is an XSD change in this PR.

File: liquibase-standard/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.21.xsd
Comment (MalloD12) [reply liquibase-standard/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.21.xsd:1]: My apologies, I haven't seen your comment before. Yeah, I agree with you. I just moved these attribute changes to the latest version. Thank you!

**Comment by MalloD12** [comment]
> Thanks for the well wishes @XDelphiGrl and @MalloD12! Everything went well and my wife and baby girl are doing great!
> 
> Was hoping to see this merged in by now though! I'm sure you guys are busy with the 4.22 issue with the new checksum algorithm, so I get it though! This does have a XSD change and so I see the releaseMajor tag was applied. Do we think this can be released with 4.23 this month? Or is there anything else that needs to be done here?

Glad to read everything went well @jasonlyle88, for some reason, I might miss Erzebet's comment. I just applied her suggestion so it's ready to be reviewed it again. 

Thanks,
Daniel.

**Review by suryaaki2 — APPROVED** [review]

File: liquibase-standard/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.21.xsd
Comment (XDelphiGrl) [reply liquibase-standard/src/main/resources/www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.21.xsd:1]: Thanks for making this change, @MalloD12.

File: liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java
Comment (XDelphiGrl) [inline liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:629]: Most `@deprecated` annotations include the method to use instead of the deprecated method. Can we add that information to this `@deprecated` as well? It might also be useful to note what should be used instead of `LabelExpression`. 

CC @MalloD12 @filipelautert

**Comment by XDelphiGrl** [comment]
@MalloD12, hi -

Thanks for including this information in the PR! It was helpful during my review.
>     * Rename attribute name from `errorIfMissingOrEmpty` to `errorIfMissing`. I'll leave this open for discussion, but I basically consider this change needed as we are not adding any extra logic to check and do some action if the change log is empty.

CC @rberezen

File: liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java
Comment (XDelphiGrl) [inline liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:757]: This `@Deprecated` should also include what method to use instead of this one.

File: liquibase-standard/src/main/java/liquibase/parser/core/yaml/YamlChangeLogParser.java
Comment (XDelphiGrl) [inline liquibase-standard/src/main/java/liquibase/parser/core/yaml/YamlChangeLogParser.java:105]: Are there other changelog parser classes (like JSON??) that would also need this update?

CC @rberezen @MalloD12

File: liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy
Comment (XDelphiGrl) [inline liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:479]: I did not even realize there is a `errorOnCircularIncludeAll` parameter! This probably should also be added to the documentation when `errorIfMissing` is added.

CC @MalloD12 @rberezen

**Review by XDelphiGrl — CHANGES_REQUESTED** [review]
I've left a few questions and suggestions in comments.

CHANGES REQUESTED

File: liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java
Comment (MalloD12) [reply liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:629]: Sure, I'll make this update.

File: liquibase-standard/src/main/java/liquibase/parser/core/yaml/YamlChangeLogParser.java
Comment (MalloD12) [reply liquibase-standard/src/main/java/liquibase/parser/core/yaml/YamlChangeLogParser.java:105]: My understanding is that there is not, because for example, JSON changelog parser extends from this one (YamlChangeLogParser), so I think with this one we are ok.

File: liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy
Comment (MalloD12) [reply liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:479]: Ups I didn't see this comment before of creating a PD ticket. I'll add a comment there so the guys can also include this when working on the created ticket.

Comment (MalloD12) [reply liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:479]: I think it's not needed. I just found it's documented here https://docs.liquibase.com/parameters/error-on-circular-include-all.html?Highlight=errorOnCircularIncludeAll

File: liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy
Comment (MalloD12) [reply liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:595]: yeah, it might sound a bit confusing and we can update it. I understood exactly what you described in your comment. I can rename it to "no error is thrown when properties file is not found and errorIfMissing" is either set as false or null. @jasonlyle88 do you agree? I'll update as I mentioned, but please feel free to make any other update on top of mine if you have a better name.

Thanks,
Daniel.

Comment (jasonlyle88) [reply liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy]: That is much clearer, thanks for the update. Amazing what makes sense when you are in the middle of something... but when you go back and read it you just question how you ever thought that was a good explanation 🤣 Thanks for the adjustment there!

File: liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java
Comment (XDelphiGrl) [reply liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:630]: Thanks!

File: liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy
Comment (XDelphiGrl) [reply liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:479]: I see; it's a global parameter and not a changeset property. My mistake. Thanks for the link to the documentation!

**Review by XDelphiGrl — APPROVED** [review]
This PR allows users to set the `errorIfMissing` attribute on the `includeAll` and `include` change types. The implementation aligns with the earlier PR to add `errorIfMissing` for the `property` changelog type (when `property` references a file, not inline key/value definitions). Adding `errorIfMissing="true"` to the `includeAll` or `include` change types prevents Liquibase from  halting execution if a referenced file is missing. This is useful in cases where it is expected for a particular file to be unavailable in a particular environment; for example, when there are `include`ed files that are required only in TEST but not in STAGING. 
* The default for `errorIfMissing` is `true`.
* Test harness execution passing.
* Functional test failures are due to a mismatch in Core/Pro code and tests that are ahead of that code. Failures are unrelated to this PR and can be disregarded.
* No additional testing required. 

In a final word, please join me on congratulating @jasonlyle88 and his wife on the birth of their new baby! @jasonlyle88 is a friendly and committed member of our Liquibase Community. That he's a new dad AND still working on OSS PRs is astounding and deserving of all the thanks in the world!

CC @rberezen

**Comment by jasonlyle88** [comment]
Thank you for your kind words @XDelphiGrl! And thank you and all the other developers here for all the work you put in on this product!

**Pull request merged**
